Mathlib/CategoryTheory/Distributive/Monoidal.lean

```diff
@@ -0,0 +1,321 @@
+- The category of pointed types with the monoidal structure given by the smash product of
+  pointed types and the coproduct given by the wedge sum.
+
+## References
+
+[Hans-Joachim Baues, Mamuka Jibladze, Andy Tonks, Cohomology of
+ monoids in monoidal categories, in: Operads: Proceedings of Renaissance
+ Conferences, Contemporary Mathematics 202, AMS (1997) 137-166][]
+
+-/
+
+universe v v₂ u u₂
+
+noncomputable section
+
+namespace CategoryTheory
+
+open Category MonoidalCategory Limits Iso
+
+/-- A monoidal category with binary coproducts is left distributive
+if the left tensor product functor preserves binary coproducts. -/
+class IsMonoidalLeftDistrib (C : Type u) [Category.{v} C]
+    [MonoidalCategory.{v} C] [HasBinaryCoproducts C] : Prop where
+  preservesBinaryCoproducts_tensorLeft (X : C) :
+    PreservesColimitsOfShape (Discrete WalkingPair) (tensorLeft X)
+
+/-- A monoidal category with binary coproducts is right distributive
+if the right tensor product functor preserves binary coproducts. -/
+class IsMonoidalRightDistrib (C : Type u) [Category.{v} C]
+    [MonoidalCategory.{v} C] [HasBinaryCoproducts C] : Prop where
+  preservesBinaryCoproducts_tensorRight (X : C) :
+    PreservesColimitsOfShape (Discrete WalkingPair) (tensorRight X)
+
+/-- A monoidal category with binary coproducts is distributive
+if it is both left and right distributive. -/
+class IsMonoidalDistrib (C : Type u) [Category.{v} C]
+    [MonoidalCategory.{v} C] [HasBinaryCoproducts C] extends
+  IsMonoidalLeftDistrib C, IsMonoidalRightDistrib C
+
+variable {C} [Category.{v} C] [MonoidalCategory.{v} C] [HasBinaryCoproducts C]
+
+example [IsMonoidalDistrib C] : IsMonoidalLeftDistrib C := by infer_instance
+
+section IsMonoidalLeftDistrib
+
+instance IsMonoidalLeftDistrib.preserves_binary_coproducts_tensorLeft
+    [IsMonoidalLeftDistrib C] {X : C} :
+    PreservesColimitsOfShape (Discrete WalkingPair) (tensorLeft X) :=
+  IsMonoidalLeftDistrib.preservesBinaryCoproducts_tensorLeft X
+
+instance IsMonoidalLeftDistrib.preservesColimit_pair_tensorLeft
+    [IsMonoidalLeftDistrib C] {X Y Z : C} :
+    PreservesColimit (pair Y Z) (tensorLeft X) :=
+  (IsMonoidalLeftDistrib.preservesBinaryCoproducts_tensorLeft X).preservesColimit
+
+/-- The canonical left distributivity isomorphism -/
+def leftDistrib [IsMonoidalLeftDistrib C] (X Y Z : C) :
+    (X ⊗ Y) ⨿ (X ⊗ Z) ≅ X ⊗ (Y ⨿ Z) :=
+  PreservesColimitPair.iso (tensorLeft X) Y Z
+
+/-- Notation for the forward direction morphism of the canonical left distributivity isomorphism -/
+scoped notation "∂L" => leftDistrib
+
+instance IsMonoidalLeftDistrib.isIso_leftDistrib_hom [IsMonoidalLeftDistrib C] {X Y Z : C} :
+    IsIso (∂L X Y Z).hom :=
+  isIso_hom <| leftDistrib X Y Z
+
+instance IsMonoidalLeftDistrib.of_isIso_coprodComparisonTensorLeft
+    [i : ∀ {X Y Z : C}, IsIso (coprodComparison (tensorLeft X) Y Z)] : IsMonoidalLeftDistrib C where
+  preservesBinaryCoproducts_tensorLeft X := by
+    refine {
```

Comment:
You are actually proving a more general (useful) lemma `preservesBinaryCoproducts_of_isIso_coprodComparison`, which should be added to `Limits.Preserves.Shapes.BinaryProducts`.